src/BackgroundPublish.h

```diff
@@ -117,78 +137,174 @@ class BackgroundPublish {
+                           event.event_name,
+                           event.event_data,
+                           event.event_context);
+    } else {
+        if (error != particle::Error::NONE) {
+            // log error if no callback is used
+            logger.error("publish failed: %s", error.message());
+        }
+    }
+
+    return error;
+}
+
+template<std::size_t NumQueues>
+void BackgroundPublish<NumQueues>::thread() {
+    constexpr std::size_t burst_rate {2u}; // allowable burst rate (Hz), Device OS allows up to 4/s
+    constexpr system_tick_t process_interval {1000u};
+
+    system_tick_t publish_t[burst_rate] {}; // publish time of the last (burst_rate) sends in a circular buffer
+    std::size_t i {}; // publish time of the previous (burst_rate)th send
+
+    while(running) {
+        auto now {millis()};
+        if(now - publish_t[i] >= process_interval) {
+            for(auto &queue : _queues) {
+                _mutex.lock();
+                if(!queue.empty()) {
+                    publish_t[i] = now;
+                    i = (i + 1) % burst_rate;
+                    // Copy the event and pop so the publish and wait is done without holding the mutex
+                    publish_event_t event {queue.front()};
+                    queue.pop();
+                    _mutex.unlock();
+                    process_publish(event);
+                    break;
+                }
+                _mutex.unlock();
+            }
+        }
+
+        delay(1); // force yield to processor
+    }
+
+    // Exit thread
+    os_thread_exit(nullptr);
+}
+
+template<std::size_t NumQueues>
+bool BackgroundPublish<NumQueues>::publish(const char *name,
+                                           const char *data,
+                                           PublishFlags flags,
+                                           std::size_t priority,
+                                           publish_completed_cb_t cb,
+                                           const void *context)
+{
+    if (!running) {
+        logger.error("publisher not initialized");
+        return false;
+    }
+
+    publish_event_t event_details{};
+    event_details.event_flags = flags;
+    event_details.completed_cb = cb;
+    event_details.event_context = context;
+
+    std::strncpy(event_details.event_name, name, sizeof(event_details.event_name));
+    event_details.event_name[sizeof(event_details.event_name) - 1] = '\0';
+    std::strncpy(event_details.event_data, data, sizeof(event_details.event_data));
+    event_details.event_data[sizeof(event_details.event_data) - 1] = '\0';
+
+    if (priority >= NumQueues) {
+        logger.error("priority %d exceeds number of queues %d", priority, NumQueues);
+        return false;
+    }
+
+    std::lock_guard<RecursiveMutex> lock(_mutex);
+
+    if(_queues[priority].size() >= maxEntries) {
+        logger.error("queue at priority %d is full");
+        return false;
+    }
+    _queues[priority].push(event_details);
```

Comment:
It would be important to state how the `event_details` payload will stay in scope long enough for the item to be popped from the queue.  The user may be trusted to keep buffers around that the eventual publish will use.  If the publish is known to happen by callback then the caller can perform `delete` or `free()` on a dynamic buffer that they allocated for this event.